### PR TITLE
:sparkles: Fortify: Support .fpr format

### PR DESCRIPTION
**Pull request by manuel-sommer**
see issue #9584

### CODE REVIEW

**Review by mtesauro — APPROVED** [review]
Approved

File: docs/content/en/integrations/parsers/file/fortify.md [hunk 1]
```diff
@@ -2,7 +2,8 @@
 title: "Fortify"
 toc_hide: true
 ---
-Import Findings from XML file format.
+You can either import the findings in .xml or in .fdr file format. </br>
```
Comment (schdief) [inline docs/content/en/integrations/parsers/file/fortify.md:5]:
```suggestion
You can either import the findings in .xml or in .fpr file format. </br>
```

File: docs/content/en/integrations/parsers/file/fortify.md [hunk 2]
```diff
@@ -2,7 +2,8 @@
...
+If you import a .fdr file, the parser will look for the file 'audit.fvdl' and analyze it. An extracted example can be found [here](https://github.com/DefectDojo/django-DefectDojo/tree/master/unittests/scans/fortify/audit.fvdl).
```
Comment (schdief) [inline docs/content/en/integrations/parsers/file/fortify.md:6]:
```suggestion
If you import a .fpr file, the parser will look for the file 'audit.fvdl' and analyze it. An extracted example can be found [here](https://github.com/DefectDojo/django-DefectDojo/tree/master/unittests/scans/fortify/audit.fvdl).
```

File: dojo/tools/fortify/parser.py [hunk 3]
```diff
@@ -18,6 +18,12 @@ def get_description_for_scan_types(self, scan_type):
         return "Import Findings from XML file format."
```
Comment (schdief) [inline dojo/tools/fortify/parser.py:18]: shouldn't this be adapted as well?

File: dojo/tools/fortify/parser.py [hunk 4]
```diff
@@ -114,6 +120,51 @@ def get_findings(self, filename, test):
...
+                    Confidence = vuln.find(f"{namespace}InstanceInfo").find(f"{namespace}Confidence").text
+                    description = Type + "\n"
+                    description += "**ClassID:** " + ClassID + "\n"
+                    description += "**Kingdom:** " + Kingdom + "\n"
+                    description += "**AnalyzerName:** " + AnalyzerName + "\n"
+                    description += "**DefaultSeverity:** " + DefaultSeverity + "\n"
+                    description += "**InstanceID:** " + InstanceID + "\n"
+                    description += "**InstanceSeverity:** " + InstanceSeverity + "\n"
+                    description += "**Confidence:** " + Confidence + "\n"
+                    items.append(
```
Comment (schdief) [inline dojo/tools/fortify/parser.py:156]: I think we should also import:
- SourceLocation path="public/category.html" as "File path" in GUI
- line="222" as "Line number" in GUI

File: dojo/tools/fortify/parser.py [hunk 5]
```diff
@@ -114,6 +120,51 @@ def get_findings(self, filename, test):
...
+                        Finding(
+                            title=Type + " " + ClassID,
+                            severity="High",
+                            static_finding=True,
+                            test=test,
+                            description=description,
+                            unique_id_from_tool=ClassID,
+                        )
+                    )
+        return items
```
Comment (schdief) [inline dojo/tools/fortify/parser.py:166]: can we also populate the test entry with data from the report like:
- Version>front_end_20240220_629_9_1361_6018341</Version>
- CreatedTS date="2024-02-20" time="05:34:11"/>
- end = created + ScanTime value="215"/>
- description based on:

```
  <NumberFiles>58</NumberFiles>
  <LOC type="Line Count">2942</LOC>
  <SourceBasePath>/var/lib/jenkins/workspace/SDM_SockShop_front-end_master/src</SourceBasePath>
  <SourceFiles> as list of files
```

File: unittests/tools/test_fortify_parser.py [hunk 6]
```diff
@@ -75,3 +74,14 @@ def test_fortify_issue6082(self):
...
+    def test_fortify_many_fdr_findings(self):
+        testfile = open("unittests/scans/fortify/many_findings.fpr")
+        parser = FortifyParser()
+        findings = parser.get_findings(testfile, Test())
+        self.assertEqual(61, len(findings))
+        with self.subTest(i=0):
+            finding = findings[0]
+            self.assertEqual("Cross-Site Request Forgery 114E5A67-3446-4DD5-B578-D0E6FDBB304E", finding.title)
+            self.assertEqual("High", finding.severity)
+            self.assertEqual('114E5A67-3446-4DD5-B578-D0E6FDBB304E', finding.unique_id_from_tool)
```
Comment (schdief) [inline unittests/tools/test_fortify_parser.py:87]: this is actually the class id and not the finding id

**Review by schdief — CHANGES_REQUESTED** [review]
great work, happy to see such a fast reaction time, added some comments for improvement

File: unittests/tools/test_fortify_parser.py (hunk 6 above)
Comment (manuel-sommer) [reply unittests/tools/test_fortify_parser.py:87]: This check is right I would say.

File: dojo/tools/fortify/parser.py (hunk 5 above)
Comment (manuel-sommer) [reply dojo/tools/fortify/parser.py:194]: I don't know if we can populate this info to the test itself. Maybe I don't understand exactly what you are talking about @schdief .

**Comment by manuel-sommer** [comment]
Please review again @schdief

File: dojo/tools/fortify/parser.py (hunk 5 above)
Comment (schdief) [reply dojo/tools/fortify/parser.py:194]: when you import a new scan result this is related to a test in defect dojo, currently you only write data for the findings

https://URL/finding/39

but it belongs to a 

https://URL/test/7

and this test has metadata itself

image

here we can add more data based on the fpr like version, start and end and within the description of the test also the number of files, loc and the list of sourcefiles from which sourcebaspath

**Review by schdief — APPROVED** [review]
lgtm, thanks again
as written, maybe we can enrich the test metadata too, to give more context about the scan, which is especially imprtant if there are no findings at all, to verify that the scan itself was valid (not just empty files scanned etc.)

what is the release cadence for defectdojo? when can we use this after merging?

File: dojo/tools/fortify/parser.py (hunk 5 above)
Comment (manuel-sommer) [reply dojo/tools/fortify/parser.py:194]: Yes, please look at the parser, I can't affect the test, only the findings. The test is affected by your API calls.

Comment (schdief) [reply dojo/tools/fortify/parser.py:194]: I can see that the current implementation cannot update the test, but the GUI uses this form
https://URL/engagement/3/import_scan_results

to import scan results, e. g. of fortify, and at the same time creates a test container for it, maybe this should be another ticket/pr, but if the info doesn't get parsed, it won't be able to be stored later on

**Comment by manuel-sommer** [comment]
> lgtm, thanks again as written, maybe we can enrich the test metadata too, to give more context about the scan, which is especially imprtant if there are no findings at all, to verify that the scan itself was valid (not just empty files scanned etc.)
> 
> what is the release cadence for defectdojo? when can we use this after merging?

https://documentation.defectdojo.com/contributing/branching-model/

**Review by devGregA — APPROVED** [review]

**Review by Maffooch — APPROVED** [review]

**Pull request merged**